source/Irrlicht/CBAWMeshFileLoader.cpp

```diff
@@ -208,10 +211,91 @@ void* CBAWMeshFileLoader::tryReadBlobOnStack(const SBlobData & _data, SContext &
 		dst = _stackPtr;
 	else
 		dst = malloc(_data.header->blobSizeDecompr);
+
+	const bool compressed = (_data.header->compressionType != core::Blob::EBCT_RAW);
+
+	uint8_t stackCompressed[1u<<14];
+	void* dstCompressed = stackCompressed; // ptr to mem to load possibly compressed data
+	if (!compressed)
+		dstCompressed = dst;
+	else if (_data.header->blobSize > sizeof(stackCompressed))
+		dstCompressed = malloc(_data.header->blobSize);
+
 	_ctx.file->seek(_data.absOffset);
-	_ctx.file->read(dst, _data.header->blobSizeDecompr);
+	_ctx.file->read(dstCompressed, _data.header->blobSize);
+
+	if (compressed)
+	{
+		const uint8_t comprType = _data.header->compressionType;
+		bool res = false;
+		switch (comprType)
+		{
+			case core::Blob::EBCT_LZ4:
+				res = decompressLz4(dst, _data.header->blobSizeDecompr, dstCompressed, _data.header->blobSize);
+				break;
+			case core::Blob::EBCT_LZMA:
+				res = decompressLzma(dst, _data.header->blobSizeDecompr, dstCompressed, _data.header->blobSize);
+				break;
+		}
+
+		if (dstCompressed != stackCompressed)
+			free(dstCompressed);
+		if (!res)
+		{
+			if (dst != _stackPtr)
+				free(dst);
+			return NULL;
+		}
+	}
 
 	return dst;
 }
 
+bool CBAWMeshFileLoader::decompressLzma(void* _dst, size_t _dstSize, const void* _src, size_t _srcSize) const
+{
+	SizeT dstSize = _dstSize;
+	SizeT srcSize = _srcSize - LZMA_PROPS_SIZE;
+	ELzmaStatus status;
+	ISzAlloc alloc{&LzmaMemMngmnt::alloc, &LzmaMemMngmnt::release};
+	const SRes res = LzmaDecode((Byte*)_dst, &dstSize, (const Byte*)(_src)+LZMA_PROPS_SIZE, &srcSize, (const Byte*)_src, LZMA_PROPS_SIZE, LZMA_FINISH_ANY, &status, &alloc);
+	if (res != SZ_OK)
+		return false;
+	return true;
+}
+
+bool CBAWMeshFileLoader::decompressLz4(void * _dst, size_t _dstSize, const void * _src, size_t _srcSize) const
+{
+	int res = LZ4_decompress_safe((const char*)_src, (char*)_dst, _srcSize, _dstSize);
+	return res >= 0;
+}
+
+void CBAWMeshFileLoader::SContext::regularDrop()
+{
+	for (std::map<uint64_t, SBlobData>::iterator it = blobs.begin(); it != blobs.end(); ++it)
+	{
+		const uint32_t type = it->second.header->blobType;
+		const uint64_t handle = it->second.header->handle;
+
+		using namespace core;
+		uint32_t typesNeededToBeDropped[]
```

Comment:
what's this about?

dont we releaseObj everything?

Reply:
It was done quite a long ago due to mem-leaks patching. Irrlicht's reference counted objects has to be dropped after setting them as other object 'ownership'. E.g.:
```
ICPUMeshBuffer* mesh = SCPUMesh();
{
ICPUMeshDataFormatDesc* desc = new ICPUMeshDataFormatDesc();
mesh->setMeshDataAndFormat(desc);
}
```
Here before `desc` goes out of scope it has to dropped so that when meshbuffer is being destroyed, `desc`'s refcount will be ==1 and meshbuffer's destructor will actually destroy it.  
But it must be done after *all* such sets (since one object can be 'owned' more then once but drop must be only one) and above applies only for reference counted objects and not even all of them... So `regularDrop` is called after all the loading. What's more.. we want to support any kind of blobs so we have to assume `releaseObj` will not do just drop but actual `delete`. That's what I was talking about while whining about this reference-counting irrlicht's thing and not some kind of `unique_ptr`.

Reply:
but why do you call releaseObj only on some blobs?

Reply:
Because not all of our ref-counted objects has to be dropped (i.e. mesh through the whole loading time has ref-count==1 because no other object grabs it) and also, in case of non-reference-counted objects releaseObj would call just `delete` and we don't want it. There's just a finite set of objects that must be dropped and infinite set of possible blobs so...

Reply:
createMesh is designed to only ever return one mesh... so thats the only thing not getting deleted.

loaded textures upon getTexture are added to the IVideoDriver texture cache and can neither be dropped or deleted, removeTexture must be called upon them... and I'm totally fine with cache pollution from unreferenced textures.

Since we have infinite set of possible blobs, every single one should have releaseObj called except for the top blobs being returned (like the root blob that we started loading).

Anything thats not being returned implicitly or explicitly by load routines should get destoroyed, because thats the very definition of a memory leak.